File: test/shapiro-wilk-normality-test.js
require('@babel/register');
const fs = require('fs');
const fetch = require('node-fetch');
const parse = require('csv-parse/lib/sync');
const d3 = require('d3');
const vector = require('../src/util/stats/vector').default;
const normality = require('../src/util/stats/normality').default;
const write = require('csv-writer'); // TODO: save data to .csv 

fetch('https://raw.githubusercontent.com/RhoInc/data-library/master/data/clinical-trials/renderer-specific/adbds.csv')
    .then(response => response.text())
    .then(text =>
        parse(
            text,
            {
                columns: true
            }
        )
    )
    .then(data => {
        const summarized = d3.nest()
            .key(d => d.TEST)
            .rollup(d => {
                const vctr = new vector.Vector(d.map(di => +di.STRESN));
                const test = normality.shapiroWilk(vctr);
                test.n = d.length;

                return test;
            })
            .entries(
                data.filter(d => d.STRESN !== '')
            )
            .map(d => {
                const datum = {
                    TEST: d.key,
                    n: d.values.n,
                    statistic: d.values.w,
                    'p.value': d.values.p,
                };

                return datum;
            });
    });

File: test/shapiro-wilk-normality-test.R
library(magrittr)
library(data.table)
library(dplyr)
library(arsenal)

# read in raw data
if (!exists('data'))
    data <- 'https://raw.githubusercontent.com/RhoInc/data-library/master/data/clinical-trials/renderer-specific/adbds.csv' %>%
        fread(
            sep = ','
        )

# run Shapiro-Wilk normality test for each test
r_output <- data %>%
    filter(
        !is.na(STRESN)
    ) %>%
    select(TEST, STRESN) %>%
    group_by(TEST) %>%
    summarize(
        n = n(),
        statistic = round(shapiro.test(STRESN)$statistic, 4),
        p.value = round(shapiro.test(STRESN)$p.value, 4)
    ) %>%
    ungroup %>%
    arrange(TEST)


# read in .js output
js_output <- 'shapiro-wilk-normality-test.csv' %>%
    fread(
        sep = ','
    ) %>%
    mutate(
        statistic = round(statistic, 4),
        p.value = round(p.value, 4)
    )

# direct output of comparison of test results to text file
con <- file('shapiro-wilk-normality-test-comparison.txt')
sink(con)
sink(con, type = 'message')

    print('----------------------------------------------------------------------------------------------------')
    print('  Comparison of Shapiro-Wilk tests between JavaScript and R')
    print('----------------------------------------------------------------------------------------------------')

    # run comparison
    comparison <- comparedf(
        r_output,
        js_output
    )
    print(summary(comparison))

    # redirect output to console
sink()
sink(type = 'message')
